cmd/gitserver/server: Remove in memory map for last fetched checks

indradhanush opened the pull request:
Instead directly read the timestamp from disk. Reading this from the disk is safe since we also read from disk before ultimately writing to the database.

## Test plan

1. Tested locally.
2. Added tests.



<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles -->

sashaostrikov commented on cmd/gitserver/server/server.go:
```diff
@@ -1031,13 +1011,44 @@ func (s *Server) handleRepoUpdate(w http.ResponseWriter, r *http.Request) {
 		var statusErr, updateErr error
 
-		if debounce(req.Repo, req.Since) {
+		lastFetched, err := repoLastFetched(dir)
```
`repoLastFetched`'s documentation says:
`This breaks on file systems that do not record mtime and if Git ever changes this undocumented behavior.`

Do you think it is a good thing to make this change if we know that it will brake on some file systems?

sashaostrikov replied:
I would move `repoLastFetched` call before the `if !repoCloned(dir) && !s.skipCloneForTests` so that we don't call it two times instead of one.

sashaostrikov replied:
never mind my second comment -- we need to read it again anyway because last fetched is updated after repo sync.

this way, I feel stronger about not implementing this change because we increase disk pressure (and add disk reading latency) by the factor of 2 in this case.

indradhanush replied:
> repoLastFetched's documentation says:
> This breaks on file systems that do not record mtime and if Git ever changes this undocumented behavior.
> 
> Do you think it is a good thing to make this change if we know that it will brake on some file systems?

Overall I'm not too worried about this at the moment because we are currently using the same function to set the last fetched time [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/gitserver/server/server.go?L1930-1948) which ultimately writes to the DB. So if this breaks in the future we will have big problems anyway. 

Maybe adding a metric to track this and an alert might be a good idea to guard against that scenario in the future though. What do you think about that?

sashaostrikov commented on cmd/gitserver/server/server.go:
```diff
@@ -1031,13 +1011,44 @@ func (s *Server) handleRepoUpdate(w http.ResponseWriter, r *http.Request) {
-		if debounce(req.Repo, req.Since) {
+		lastFetched, err := repoLastFetched(dir)
+		if err != nil {
+			// TODO: Use level Info
```
Should this TODO be removed? I think `warn` is better here

sashaostrikov commented on cmd/gitserver/server/server.go:
```diff
@@ -1031,13 +1011,44 @@ func (s *Server) handleRepoUpdate(w http.ResponseWriter, r *http.Request) {
+		lastFetched, err := repoLastFetched(dir)
+		if err != nil {
+			// TODO: Use level Info
+			s.Logger.Warn(
```
let's use scoped `logger` which is defined in the beginning of the function

sashaostrikov commented on cmd/gitserver/server/server.go:
```diff
@@ -1031,13 +1011,44 @@ func (s *Server) handleRepoUpdate(w http.ResponseWriter, r *http.Request) {
+			)
+		}
+
+		// Perform a fresh repo update only iff the current time is beyond the threshold
```
lil typo fix

```suggestion
		// Perform a fresh repo update only if the current time is beyond the threshold
```

ryanslade replied:
This may have been two different typos :) 

https://en.wiktionary.org/wiki/iff

Should change either to what you suggested or leave "iff" and remove "only"

indradhanush replied:
Yep, was going for `iff` here. Should remove the preceding `only`. 👍

sashaostrikov commented on cmd/gitserver/server/server.go:
```diff
@@ -1031,13 +1011,44 @@ func (s *Server) handleRepoUpdate(w http.ResponseWriter, r *http.Request) {
+		threshold := lastFetched.Add(req.Since)
+		now := time.Now()
+		if now.After(threshold) {
+			s.Logger.Info(
```
this log will be spammy, I wouldn't add it

sashaostrikov commented on cmd/gitserver/server/server.go:
```diff
@@ -1031,13 +1011,44 @@ func (s *Server) handleRepoUpdate(w http.ResponseWriter, r *http.Request) {
+			)
 			updateErr = s.doRepoUpdate(ctx, req.Repo, "")
+		} else {
+			s.Logger.Warn("skipping repo update",
```
this log will be spammy, I wouldn't add it

sashaostrikov commented on cmd/gitserver/server/server.go:
```diff
@@ -1031,13 +1011,44 @@ func (s *Server) handleRepoUpdate(w http.ResponseWriter, r *http.Request) {
+		// Perform a fresh repo update only iff the current time is beyond the threshold
+		// of the duration in the request. This provides some filtering against spammy
+		// repo update requests.
+		threshold := lastFetched.Add(req.Since)
```
when there is an error during getting `lastFetched`, we can get a zero-valued `Time`. in this case we won't update the repo, is it an expected behaviour?

sashaostrikov commented on cmd/gitserver/server/server_test.go:
```diff
@@ -971,6 +971,39 @@ func TestHandleRepoUpdate(t *testing.T) {
 		t.Fatal(diff)
 	}
 
+	// Create a new request with the Since interval to a long duration to verify that the
```
```suggestion
	// Create a new request with the Since interval set to a long duration to verify that the
```

probably?

sashaostrikov commented on cmd/gitserver/server/server_test.go:
```diff
@@ -971,6 +971,39 @@ func TestHandleRepoUpdate(t *testing.T) {
+		Since: time.Duration(time.Hour * 1),
+	}
+
+	reqWithSince.Since = time.Duration(time.Hour * 1)
```
this looks redundant, we have already set this value above

indradhanush replied:
Left over after I edited the code around the test. Forgot to fix this. 🤦‍♂️

sashaostrikov commented on cmd/gitserver/server/server_test.go:
```diff
@@ -971,6 +971,39 @@ func TestHandleRepoUpdate(t *testing.T) {
+	}
+
+	// Reset the Since interval for further tests.
+	updateReq.Since = time.Duration(time.Second * 0)
```
this value doesn't seem to be changed by your test 🤔

indradhanush replied:
Oops. I was mutating `updateReq` earlier in the test and then edited it. Forgot to remove this bit after that. 🤦🏽‍♂️

sashaostrikov commented on cmd/gitserver/server/server_test.go:
```diff
@@ -971,6 +971,39 @@ func TestHandleRepoUpdate(t *testing.T) {
+		t.Fatal(err)
+	}
+
+	// We will not ignore any fields in the comparison because we don't expect any changes
```
alternative version suggestion:

```suggestion
	// We compare all fields because we don't expect any changes
```

sashaostrikov reviewed (commented):
I have a question about an overall idea, should this change be implemented if it can be broken on some file systems? 🤔

mrnugget commented:
Wait, what's the reason for removing it? The cache is there so we don't have to hit the file system, right? What changed that makes reading from the filesystem okay now?

I think there can be requests in which we call this multiple times. Going out to disk seems a bit hard.

indradhanush commented:
> Wait, what's the reason for removing it? The cache is there so we don't have to hit the file system, right? What changed that makes reading from the filesystem okay now?
> 
> I think there can be requests in which we call this multiple times. Going out to disk seems a bit hard.

Main motivation was to reduce complexity and remove an optimisation that we probably don't need. 

Referring to [Latency Numbers Every Programmer Should Know](https://gist.github.com/jboner/2841832), if a main memory read is 150ns, the read from disk is 150us. Are we anticipating this to be a performance hit worth optimizing for? 🤔

mrnugget commented:
> Referring to [Latency Numbers Every Programmer Should Know](https://gist.github.com/jboner/2841832), if a main memory read is 150ns, the read from disk is 150us. Are we anticipating this to be a performance hit worth optimizing for? 🤔

I'd be curious to see how often this memory map is hit when you browse a repository and run a few searches.

I'm all for reducing complexity and removing stuff we don't need, but in this case we're talking about 25 lines of code.

indradhanush commented:
I added a metric and tested locally:

![image](https://user-images.githubusercontent.com/2682729/201346995-30bb09f3-fa7a-4dbb-89dd-a523790868bf.png)

It is accessed once on site startup and once again each time I triggered a repo update. Searching for the repo or visiting it had no effect as expected.

The metric I added is available here: https://github.com/sourcegraph/sourcegraph/commit/61e155182fc11df4826797a94c59bd00b6d164e5

Also, I benchmarked the code for accessing the map locked with a mutex vs reading the timestamp from disk, and expectedly accessing the map is faster. But the performance hit from reading from disk is negligible:

```
➜ go test -bench .
goos: darwin
goarch: arm64
pkg: github.com/sourcegraph/sourcegraph/benchmark-test
BenchmarkDebounce-10           	10133280	       117.8 ns/op
BenchmarkRepoLastFetched-10    	  459228	      2629 ns/op
PASS
ok  	github.com/sourcegraph/sourcegraph/benchmark-test	3.155s
```

The benchmarking code is available here: https://github.com/sourcegraph/sourcegraph/commit/682b352bf47b25cdcb3bce845abb1b77db0fff08

Overall, my feeling towards using a separate map for tracking this state as an independent source of truth is a code smell and potentially a good place for bugs in the future (although in its current form the piece of code is quite simple). And we should avoid multiple sources of truth where possible.

Looking at it from the other way round, if we already had the code from this PR in `main`, would we feel comfortable accepting a PR that tracked the `lastFetched` time in an independent map for a minor performance gain?

However, on the balance of things if we still don't feel right about going ahead with dropping the in memory map, I'll be happy to close the PR.

ryanslade commented:
I also think that it might not be worth making this change. As @mrnugget mentioned, it's a pretty small refactor and isn't risk free. It's also not code we need to touch often so feels safe to leave as it is.
